library/init/category/state.lean

```diff
@@ -1,124 +1,131 @@
+structure state_t (σ : Type u) (m : Type u → Type v) (α : Type u) : Type (max u v) :=
+(run' : σ → m (α × σ))
 
-namespace state
-@[inline] def read {σ : Type u} : state σ σ :=
-λ s, (s, s)
+attribute [pp_using_anonymous_constructor] state_t
 
-@[inline] def write {σ : Type} : σ → state σ unit :=
-λ s' s, ((), s')
+@[reducible] def state (σ α : Type u) : Type u := state_t σ id α
 
-@[inline] def write' {σ : Type u} : σ → state σ punit :=
-λ s' s, (punit.star, s')
-end state
+namespace state_t
+section
+  variables {σ : Type u} {m : Type u → Type v}
 
-def state_t (σ : Type u) (m : Type u → Type v) [monad m] (α : Type u) : Type (max u v) :=
-σ → m (α × σ)
+  -- swap `st` and `x`
+  @[inline] protected def run {α : Type u} (st : σ) (x : state_t σ m α) : m (α × σ) :=
+  state_t.run' x st
 
-section
-  variable  {σ : Type u}
-  variable  {m : Type u → Type v}
   variable  [monad m]
   variables {α β : Type u}
 
-  def state_t_return (a : α) : state_t σ m α :=
-  λ s, show m (α × σ), from
-    return (a, s)
+  @[inline] protected def pure (a : α) : state_t σ m α :=
+  ⟨λ s, pure (a, s)⟩
+
+  @[inline] protected def bind (x : state_t σ m α) (f : α → state_t σ m β) : state_t σ m β :=
+  ⟨λ s, do (a, s') ← x.run s,
+           (f a).run s'⟩
+
+  instance : monad (state_t σ m) :=
+  { pure := @state_t.pure _ _ _, bind := @state_t.bind _ _ _ }
+
+  protected def orelse [alternative m] {α : Type u} (x₁ x₂ : state_t σ m α) : state_t σ m α :=
+  ⟨λ s, x₁.run s <|> x₂.run s⟩
+
+  protected def failure [alternative m] {α : Type u} : state_t σ m α :=
+  ⟨λ s, failure⟩
+
+  instance [alternative m] : alternative (state_t σ m) :=
+  { failure := @state_t.failure _ _ _ _,
+    orelse  := @state_t.orelse _ _ _ _ }
+
+  @[inline] protected def get : state_t σ m σ :=
+  ⟨λ s, pure (s, s)⟩
+
+  @[inline] protected def put : σ → state_t σ m punit :=
+  λ s', ⟨λ s, pure (punit.star, s')⟩
+
+  @[inline] protected def modify (f : σ → σ) : state_t σ m punit :=
+  ⟨λ s, pure (punit.star, f s)⟩
 
-  def state_t_bind (act₁ : state_t σ m α) (act₂ : α → state_t σ m β) : state_t σ m β :=
-  λ s, show m (β × σ), from
-     do (a, new_s) ← act₁ s,
-        act₂ a new_s
+  @[inline] protected def lift {α : Type u} (t : m α) : state_t σ m α :=
+  ⟨λ s, do a ← t, pure (a, s)⟩
+
+  instance : has_monad_lift m (state_t σ m) :=
+  ⟨@state_t.lift σ m _⟩
+
+  @[inline] protected def monad_map {σ m m'} [monad m] [monad m'] {α} (f : Π {α}, m α → m' α) :
+    state_t σ m α → state_t σ m' α :=
+  λ x, ⟨λ st, f (x.run st)⟩
+
+  instance (σ m m') [monad m] [monad m'] : monad_functor m m' (state_t σ m) (state_t σ m') :=
+  ⟨@state_t.monad_map σ m m' _ _⟩
+
+  -- TODO(Sebastian): uses lenses as in https://hackage.haskell.org/package/lens-4.15.4/docs/Control-Lens-Zoom.html#t:Zoom ?
+  protected def zoom {σ σ' α : Type u} {m : Type u → Type v} [monad m] (get : σ → σ')
+    (set : σ' → σ → σ) (x : state_t σ' m α) : state_t σ m α :=
+  ⟨λ st, (λ p : α × σ', (p.fst, set p.snd st)) <$> x.run (get st)⟩
```

Comment:
Could you please replace this definition with 
```
  ⟨λ st, do (a, st') ← x.run' (get st),
            pure (a, set st' st)⟩
```
?
BTW, it would be nice to give examples why `zoom` is useful.
I think it may produce performance problems since the state is not being used linearly.
For example, suppose `st` is an array, then when executing `x` and `get` we will not be able to perform destructive updates.